plugin-server/src/worker/ingestion/event-pipeline/populateTeamDataStep.ts

```diff
@@ -1,28 +1,80 @@
 import { PluginEvent } from '@posthog/plugin-scaffold'
+import { Counter } from 'prom-client'
 
+import { eventDroppedCounter } from '../../../main/ingestion-queues/metrics'
 import { PipelineEvent } from '../../../types'
 import { EventPipelineRunner } from './runner'
 
-/* 
-This step populates event.team_id and deletes event.ip if needed.
-If the event already has a team_id we will not run this step and 
-the capture endpoint will have handled this process. This is 
-temporary as this step will become the default for all events
-when we fully remove this functionality from the capture endpoint.
-*/
+export const inconsistentTeamCounter = new Counter({
+    name: 'ingestion_inconsistent_team_resolution_total',
+    help: 'Count of events with an inconsistent team resolution between capture and plugin-server, should be zero.',
+})
+
+export const tokenOrTeamPresentCounter = new Counter({
+    name: 'ingestion_event_hasauthinfo_total',
+    help: 'Count of events by presence of the team_id and token field.',
+    labelNames: ['team_id_present', 'token_present'],
+})
+
 export async function populateTeamDataStep(
     runner: EventPipelineRunner,
     event: PipelineEvent
 ): Promise<PluginEvent | null> {
+    /**
+     * Implements team_id resolution and applies the team's ingestion settings (dropping event.ip if requested).
+     *
+     * If the event already has a team_id field set by capture, it is used, but plugin-server still runs
+     * the resolution logic to confirm no inconsistency exists. Once team_id resolution is fully removed
+     * from capture, that section should be resolved, and team_id not trusted anymore.
+     */
+
+    // Collect statistics on the shape of events we are ingesting
+    tokenOrTeamPresentCounter
+        .labels({
+            team_id_present: event.team_id ? 'true' : 'false',
+            token_present: event.token ? 'true' : 'false',
+        })
+        .inc()
+
+    // If a team_id is present (resolved at capture, or injected by an app), trust it but
+    // try resolving from the token if present, and compare results to detect inconsistencies.
+    // TODO: remove after lightweight capture is fully rolled-out and the
+    //       ingestion_event_hasauthinfo_total metric confirms all incoming events have a token.
+    if (event.team_id) {
+        if (event.token) {
+            const team = await runner.hub.teamManager.getTeamByToken(event.token)
+            if (team?.id || event.team_id) {
+                inconsistentTeamCounter.inc()
+                runner.hub.statsd?.increment('ingestion_inconsistent_team_resolution_total')
+            }
+        }
```

Comment:
statsd the else case here too

Reply:
Changing the prom counter to a `check_ok:true/false` tag, and removing the statsd dupe, as I'm pretty sure we're in the main loop 🤞